### PR TITLE
Add proper equality for TableSection

### PR DESCRIPTION
**Pull request by abecevello**
Currently TableSection equality conformance only checks if the keys are equal. Although this specific equality function isn't used during FTD diffing, the fact this equality function is incomplete is causing sections to fail to update in some limited cases.

This PR implements proper equality of TableSections by checking all properties, including all rows.

### CODE REVIEW

File: FunctionalTableData/TableSection.swift [hunk 1]
```diff
@@ -164,6 +164,20 @@ extension Array where Element: TableSectionType {
 
 extension TableSection: Equatable {
 	public static func ==(lhs: TableSection, rhs: TableSection) -> Bool {
-		return lhs.key == rhs.key
+		return lhs.key == rhs.key &&
+			(lhs.header == nil && rhs.header == nil || lhs.header?.isEqual(rhs.header) ?? false) &&
```
Comment (raulriera) [inline FunctionalTableData/TableSection.swift]: Why are you checking for the nil? Wouldn't isEqual do this too?

File: FunctionalTableData/TableSection.swift [hunk 2]
```diff
@@ -164,6 +164,20 @@ extension Array where Element: TableSectionType {
...
+			(lhs.footer == nil && rhs.footer == nil || lhs.footer?.isEqual(rhs.footer) ?? false) &&
+			isEqual(lhs: lhs.rows, rhs: rhs.rows) &&
```
Comment (raulriera) [inline FunctionalTableData/TableSection.swift]: The equality of the rows happen somewhere else, no need to do this here right?

File: FunctionalTableData/TableSection.swift [hunk 3]
```diff
@@ -164,6 +164,20 @@ extension Array where Element: TableSectionType {
...
+			lhs.style == rhs.style &&
+			(lhs.headerVisibilityAction == nil && rhs.headerVisibilityAction == nil || lhs.headerVisibilityAction != nil && rhs.headerVisibilityAction != nil) &&
+			(lhs.didMoveRow == nil && rhs.didMoveRow == nil || lhs.didMoveRow != nil && rhs.didMoveRow != nil)
```
Comment (raulriera) [inline FunctionalTableData/TableSection.swift]: You shouldn't include the closures in the equality checks

File: FunctionalTableData/TableSection.swift (hunk 1 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: `TableHeaderFooterConfigType` is not equatable but has a `isEqual` method defined for it. This means the `isEqual` method call has to handle nil here.

File: FunctionalTableData/TableSection.swift (hunk 2 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: The `==` function is only used in 2 places in a closed source framework built on this framework. In order to ensure this equality function as defined remains correct, it needs to do row equality checks here. You are correct that in normal operation the diffing algorithm handles row comparisons separately outside of this equality function, but I think that might just be due to differing times that the diffing and this function were created.

File: FunctionalTableData/TableSection.swift (hunk 1 above)
Comment (raulriera) [reply FunctionalTableData/TableSection.swift]: What I mean is why not just use this

`lhs.header?.isEqual(rhs.header)`. Wouldn't isEqual return false if one of them is `nil`?

File: FunctionalTableData/TableSection.swift (hunk 3 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: @g-Off said to include them like this, but I can remove them.

File: FunctionalTableData/TableSection.swift (hunk 2 above)
Comment (raulriera) [reply FunctionalTableData/TableSection.swift]: Wouldn't this essentially run them twice? Equality check on all rows when checking the TableSection equality and then equality check on all rows in `TableSectionChangeSet`?

File: FunctionalTableData/TableSection.swift (hunk 1 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: If one of them is nil and the other isn't then `isEqual` should return false, but if both are nil then `isEqual` should return true. That line of code you suggest doesn't handle that case.

File: FunctionalTableData/TableSection.swift (hunk 2 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: I ran the code and this `==` function isn't run during the diffing algorithm. I commented out this entire function and it was only used in 2 places, none of which were in the FunctionalTableData framework.

Comment (raulriera) [reply FunctionalTableData/TableSection.swift]: 😳

File: FunctionalTableData/TableSection.swift (hunk 3 above)
Comment (g-Off) [reply FunctionalTableData/TableSection.swift]: A nil `didMoveRow` to non-nil, or vice-versa, should could as not being equal (same for the `headerVisibilityAction`)

**Comment by g-Off** [comment]
Can we add a couple of unit tests as well?

File: FunctionalTableData/TableSection.swift (hunk 3 above)
Comment (abecevello) [reply FunctionalTableData/TableSection.swift]: Added back.

**Comment by abecevello** [comment]
@g-off Tests have been added

**Pull request merged**